### ISSUE TITLE
Clearer errors that pinpoint exact issue without extraneous errors

### ISSUE THREAD

**Issue by cbautista1002** [stoplightio/spectral#221]
### **I'm submitting a...**
  - [ ] bug report
  - [x] feature request

### What is the current behavior?
When the schema is invalid, multiple seemingly-unrelated errors are returned along with the true error. These extra errors make a bit harder to pinpoint the actual issue.

### What is the expected behavior?
Only the actual error / root cause is listed as an error.

### What is the motivation / use case for changing the behavior?
Faster debugging of errors returned by Spectral. Fellow engineers would be able to more quickly fix issues in their schemas if the one actual error is provided.

### Please tell us about your environment:
```
$ spectral --version
@stoplight/spectral/3.0.0 linux-x64 node-v12.2.0
```

### Other information

Here's a sample schema:
```yaml
openapi: "3.0.0"
info:
  version: 1.0.0
  title: Swagger Petstore
  description: Swagger Petstore
  license:
    name: MIT
  contact:
    name: Test
servers:
  - url: http://petstore.swagger.io/v1
paths:
  /pets/{petId}:
    get:
      summary: Info for a specific pet
      operationId: showPetById
      description: Get a pet
      tags:
        - pets
      parameters:
        - name: petId
          in: path
          required: true
          description: The id of the pet to retrieve
          schema:
            type: string
      responses:
        '200':
          description: Expected response to a valid request
          content:
            application/json:
              schema:
                type: array
                additionalItems: false
                items:
                  required: []
                  properties:
                    id:
                      type: integer
```

There are two issues with this schema:
1. `additionalItems` is not valid OpenAPI
2. `required` cannot be an empty list

Here is what spectral returns:
```
$ spectral lint myschema.yml
Adding OpenAPI 3.x functions
OpenAPI 3.x detected

myschema.yml
 32:22  error  oas3-schema  should NOT have additional properties: additionalItems
 32:22  error  oas3-schema  should have required property '$ref'
 32:22  error  oas3-schema  should match exactly one schema in oneOf

✖ 3 problems (3 errors, 0 warnings, 0 infos)
```

In this case, the first error is the only one that from my perspective is relevant and makes sense.

If that's fixed (remove `additionalItems`), then: the error is
```
$ spectral lint myschema.yml
Adding OpenAPI 3.x functions
OpenAPI 3.x detected

myschema.yml
 34:23  error  oas3-schema  should have required property '$ref'
 34:23  error  oas3-schema  should match exactly one schema in oneOf
 35:28  error  oas3-schema  should NOT have fewer than 1 items

✖ 3 problems (3 errors, 0 warnings, 0 infos)
```

In this case, the last listed error is the "real" error. As a first time user, this was confusing. The first two listed errors sent me chasing something that wasn't there.

Great project, btw!

**Comment by P0lip** [comment]
Hey, @cbautista1002.
First of all, it's very nice to see you here and we are glad you like Spectral.
Recently, we've improved the error messages oas2-schema and oas3-schema rules report. If you are interested, you can check out the changes in greater detail in the following PR https://github.com/stoplightio/spectral/pull/215
It still needs some further work, mostly around prioritization this time.
I believe we could consider a few less important errors as warnings and perhaps ignore a few of them entirely.
There is a number of errors oas2-schema and oas3-schema rules report, so we'd need to carefully verify each and every error to make sure we don't omit any important feedback.
I cannot promise any date of when this becomes a thing, but I can assure you we won't ignore the issue.
I'd appreciate a list of errors you'd consider as warnings.
Moreover, it's totally possible for you to work on this issue and submit a PR. Don't feel obliged, but if you decide to work on it, we will certainly help you out if you are stuck. :)
Thanks again for reaching out.

**Comment by cbautista1002** [comment]
> I can assure you we won't ignore the issue.

Awesome, thanks!

For the time being I can get used to detecting which error is likely to be the root error, and I can tell my coworkers the same. (we're adding spectral to our CI) 

> If you are interested, you can check out the changes in greater detail in the following PR #215

Great improvement!

> I'd appreciate a list of errors you'd consider as warnings.

I can provide more feedback as we use spectral more, but for now I feel like the following shouldn't even be reported as warnings:
`should have required property '$ref'`
`should match exactly one schema in oneOf`
at least in the examples I provide above.

**Comment by philsturgeon** [comment]
@cbautista1002 hey, just to let you know this hasn't fallen through the cracks. We're investigating integrating better-ajv-errors so we can defer the de-duping and refining of errors to that, instead of hand rolling our own solution.
